napalm_ios/ios.py

```diff
@@ -1038,86 +1038,52 @@ def get_interfaces_ip(self):
-                                val = {'prefix_length': int(subnet)}
-                            except ValueError:
-                                val = {'prefix_length': u'N/A'}
-                            interfaces[interface]['ipv4'][ip_address] = val
-                    elif len(fields) in [4, 5]:
-                        # Check for 'ip address 10.10.10.1 255.255.255.0'
-                        # Check for 'ip address 10.10.11.1 255.255.255.0 secondary'
-                        if 'ipv4' not in interfaces[interface].keys():
-                            interfaces[interface].update({'ipv4': {}})
-                        ip_address = fields[2]
-
-                        try:
-                            subnet = sum([bin(int(x)).count('1') for x in fields[3].split('.')])
-                        except ValueError:
-                            subnet = u'N/A'
-
-                        ip_dict = {'prefix_length': subnet}
-                        interfaces[interface]['ipv4'].update({ip_address: {}})
-                        interfaces[interface]['ipv4'][ip_address].update(ip_dict)
-                    else:
-                        raise ValueError(u"Unexpected Response from the device")
-
-                # Check IPv6
-                if 'ipv6 address ' in line:
-                    fields = line.split()
-                    ip_address = fields[2]
-                    if 'ipv6' not in interfaces[interface].keys():
-                            interfaces[interface].update({'ipv6': {}})
-
-                    try:
-                        if r'/' in ip_address:
-                            # check for 'ipv6 address 1::1/64'
-                            ip_address, subnet = ip_address.split(r'/')
-                            interfaces[interface]['ipv6'].update({ip_address: {}})
-                            ip_dict = {'prefix_length': int(subnet)}
-                        else:
-                            # check for 'ipv6 address FE80::3 link-local'
-                            interfaces[interface]['ipv6'].update({ip_address: {}})
-                            ip_dict = {'prefix_length': u'N/A'}
-
-                        interfaces[interface]['ipv6'][ip_address].update(ip_dict)
-                    except AttributeError:
-                        raise ValueError(u"Unexpected Response from the device")
-
-        # remove keys with no data
-        new_interfaces = {}
-        for k, val in interfaces.items():
-            if val:
-                new_interfaces[k] = val
-        return new_interfaces
+            if(line[0] != ' '):
+                ipv4 = {}
+                interfaces[line.split()[0]] = {'ipv4': ipv4}
+            m = re.match(INTERNET_ADDRESS, line)
+            if m:
+                ip, prefix = m.groups()
+                ipv4.update({ip: {"prefix_length": int(prefix)}})
+
+        # Remove interfaces without ip
+        interfaces = dict(filter(lambda x: x[1]['ipv4'] != {}, interfaces.items()))
```

Comment:
Might this be more readable as
```python
interfaces = {intf: details for intf, details in interfaces.items() if details['ipv4'] != {}}
```
?

Reply:
Yes, I think I would probably change this slightly to something like:

```
            if(line[0] != ' '):
                ipv4 = {}
                interface_name = line.split()[0]
            m = re.match(INTERNET_ADDRESS, line)
            if m:
                ip, prefix = m.groups()
                ipv4.update({ip: {"prefix_length": int(prefix)}})
                interfaces[interface_name] = {'ipv4': ipv4}
```

i.e. only add it to 'interfaces' if there is a positive match